[Streams] [Streamlang] Enable nested conditionals in the UI

Kerry350 opened the pull request:
## Summary

Closes https://github.com/elastic/streams-program/issues/445

This PR adds support for nested conditionals (nested `where` blocks in Streamlang DSL) to the processing UI. There have been architectural changes to the state machines, and UI changes to match the [Figma](https://www.figma.com/design/C4o0y9Knk4jYXjsulvbW3w/Stream-Management--Processors--Grok?node-id=6004-42432&p=f&t=QSrMxFyNLwrO2mLq-0) designs.

## States

There are quite a lot of states within the UI, I've tried to capture the majority of them here. Also be mindful of UX additions such as scrolling newly added steps into view. 

<details>
<summary>Screenshots</summary>
<img width="1186" height="1216" alt="Screenshot 2025-09-11 at 22 17 48" src="https://github.com/user-attachments/assets/5ea4e45f-4398-4463-b5e5-81d8be005555" />
<img width="1164" height="1182" alt="Screenshot 2025-09-11 at 22 18 40" src="https://github.com/user-attachments/assets/a5604248-926c-49ec-bd00-18cae7a8aece" />
<img width="1174" height="508" alt="Screenshot 2025-09-11 at 22 18 49" src="https://github.com/user-attachments/assets/fbe19f40-761b-46fe-a2ac-6ad2a3795043" />
<img width="1150" height="246" alt="Screenshot 2025-09-11 at 22 19 05" src="https://github.com/user-attachments/assets/75f80061-985e-4600-871a-abda65d2f0d0" />
<img width="1148" height="1062" alt="Screenshot 2025-09-11 at 22 19 43" src="https://github.com/user-attachments/assets/78353291-809e-47a7-bb1c-76995559c38b" />
<img width="1176" height="664" alt="Screenshot 2025-09-11 at 22 20 10" src="https://github.com/user-attachments/assets/a92feeac-d0b4-43a8-8e44-42f34aea3a24" />
<img width="1260" height="380" alt="Screenshot 2025-09-11 at 22 20 17" src="https://github.com/user-attachments/assets/b9063711-58a0-41c1-8644-c876909734a6" />
<img width="1166" height="874" alt="Screenshot 2025-09-11 at 22 20 31" src="https://github.com/user-attachments/assets/7f593938-e447-4e68-a260-3bf39ff6af8f" />
</details>


## On the way

- ~Existing Scout tests are still being updated (this doesn't need to block reviews)~
- Update Scout tests to cover new features
- Copy needs review

Kerry350 committed:
Add support for nested conditionals to the processing UI

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/capture_oas_snapshot   --include-path /api/status   --include-path /api/alerting/rule/   --include-path /api/alerting/rules   --include-path /api/actions   --include-path /api/security/role   --include-path /api/spaces   --include-path /api/streams   --include-path /api/fleet   --include-path /api/saved_objects/_import   --include-path /api/saved_objects/_export   --include-path /api/maintenance_window --update'

kibanamachine committed:
[CI] Auto-commit changed files from 'make api-docs'

elasticmachine commented:
Pinging @elastic/obs-ux-logs-team (Team:obs-ux-logs)

Kerry350 committed:
Merge branch 'main' into 445-nested-conditionals

Kerry350 committed:
Merge branch 'main' into 445-nested-conditionals

tonyghiani commented:
Made a first pass on the functionalities and still have to check the edge cases, but overall looks great!

I found some small things that need fixes/adjustments:

- When the user does not have enough privileges, "Create your first step" is correctly disabled, but no tooltip exists, as it used to be for the previous create button. Can we restore that tooltip on button hover in this scenario?
- Condition editor is already failing before being really touched, this should ideally trigger the error when clicking the “Create condition” button if the condition is invalid.

https://github.com/user-attachments/assets/ebdbf2fd-09b7-4747-864f-63315b6687ab

- The delete processor button should be less prominent by the linked design, it should be left-aligned and take only the minimum space.

<img width="554" height="274" alt="Screenshot 2025-09-16 at 11 24 15" src="https://github.com/user-attachments/assets/0cc5a754-6a0e-4459-8940-54c8d157a382" />

- When deleting a staged processor, the simulation does not update anymore and keep the previous state with the parsed fields.

https://github.com/user-attachments/assets/72590f40-8fec-4011-a5f5-c3c9689ac2b8

- In alignment with design, it would be great to add the thin border around step panels to make them more distinguished on bright screens.

| Before | After |
|--------|--------|
|  <img width="2575" height="1208" alt="Screenshot 2025-09-16 at 11 39 53" src="https://github.com/user-attachments/assets/5e01955c-e078-4ca6-b2d5-f57fc66a02e2" /> | <img width="2576" height="1216" alt="Screenshot 2025-09-16 at 11 39 23" src="https://github.com/user-attachments/assets/09cdb4f0-be57-4c82-8b79-e48aa1a3c6c0" /> |

tonyghiani commented on x-pack/platform/plugins/shared/streams_app/public/components/data_management/stream_detail_enrichment/steps/blocks/where/index.tsx:
```diff
@@ -0,0 +1,176 @@
+                border: 1px solid ${euiTheme.colors.borderStrongPrimary};
+                box-sizing: border-box;
+              `
+            : undefined
```
**nit**: you can get the border for the where panels replacing here undefined with
```tsx
css`
  border: ${euiTheme.border.thin};
`
```

Kerry350 committed:
(Partial) PR feedback

- Change default condition to ALWAYS condition (stops immediate form error)
- Put border on panels
- Amend delete processor button width

Kerry350 committed:
Merge branch '445-nested-conditionals' of github.com:Kerry350/kibana into 445-nested-conditionals

tonyghiani reviewed (commented):
Finished a first round on the code as well, it was a very nice read 👏 
I left minor nits around, no major blockers, let me know if you have any question!

Kerry350 committed:
PR amendments: Fix text truncation on long condition summaries

Kerry350 committed:
Merge remote-tracking branch 'upstream/main' into 445-nested-conditionals

Kerry350 commented on x-pack/platform/plugins/shared/streams_app/public/components/data_management/stream_detail_enrichment/state_management/steps_state_machine/utils.ts:
```diff
@@ -0,0 +1,16 @@
+
+import type { StepActorSnapshot } from './steps_state_machine';
+
+export const isStepUnderEdit = (stepSnapshot: StepActorSnapshot) => {
+  return stepSnapshot.matches('draft') || stepSnapshot.matches({ configured: 'editing' });
+};
+
+export const isRootStep = (stepSnapshot: StepActorSnapshot) => {
+  return stepSnapshot.context.step.parentId === null;
+};
```
I'd argue this aligns with the current setup. The selectors in `selectors.ts` select / pull data, and then the functions in `utils.ts` perform utility functions. In this case these functions aren't selecting any data, they're just acting on the snapshot. This seems to align with the split we've had so far across the machines.

Kerry350 committed:
PR amendments

- Remove duplicate event types
- Improve UI step conversion functions

Kerry350 commented on x-pack/platform/packages/shared/kbn-streamlang/types/streamlang.ts:
```diff
@@ -28,20 +28,28 @@ export type ConditionWithSteps = Condition & { steps: StreamlangStep[] };
  * Nested where block (recursive)
  */
 export interface StreamlangWhereBlock {
+  customIdentifier?: string;
```
Ah, sorry, I think you're just referencing the fact it's duplicated. This is just because I want the property to be enforced and not optional on UI steps.

Kerry350 commented on x-pack/platform/plugins/shared/streams_app/public/components/data_management/stream_detail_enrichment/state_management/stream_enrichment_state_machine/stream_enrichment_state_machine.ts:
```diff
@@ -90,50 +93,96 @@ export const streamEnrichmentMachine = setup({
+        }
+      ) => {
+        if (!condition) {
+          condition = {
```
This was changed in https://github.com/elastic/kibana/pull/232322/commits/73f325bbf91566333d30c5512e9fd0de80a4de07 to use the `ALWAYS` constant as a side effect of fixing something else 👍

Kerry350 commented on x-pack/platform/plugins/shared/streams_app/public/components/data_management/stream_detail_enrichment/state_management/stream_enrichment_state_machine/stream_enrichment_state_machine.ts:
```diff
@@ -185,34 +234,28 @@ export const streamEnrichmentMachine = setup({
-          (processorRef, pos) => initialProcessorsRefs[pos]?.id !== processorRef.id
-        )
+        // Step order changed
+        stepRefs.some((stepRef, pos) => initialStepRefs[pos]?.id !== stepRef.id)
```
Yep, exactly. I've tried to model things (like very strict insertion indexes etc) on the fact DnD will be coming back.

Kerry350 committed:
PR amendments

Kerry350 committed:
PR amendment: Ensure step event is forwarded from the idle state (context menu)

Kerry350 committed:
Don't persist custom identifiers

Kerry350 commented:
@tonyghiani Thanks for the thorough reviewing 🙏

I think everything is addressed now. I've updated https://github.com/elastic/kibana/pull/232322#issuecomment-3297870347 with ✅ statuses and notes. And addressed the inline feedback. Just a couple of notes:

> When the user does not have enough privileges, "Create your first step" is correctly disabled, but no tooltip exists, as it used to be for the previous create button. Can we restore that tooltip on button hover in this scenario?

Left this for now because, unless I'm missing something, this is no different to `main`. We show an insufficient privileges tooltip for the save button in the management bar, but the add processor button is just disabled.

> It might make sense to make these functions (addCondition, addProcessor) take a single object with the params it takes, having it like this and having to pass undefined makes it a bit difficult to follow on.

I'll take a look in a followup. I really just wanted to keep the UI properties and the "pure" definition separate. 

Copy amendments can come in a followup.

Scout test expansion can also come in a followup.

I'm very cognizant of the fact this is blocking several issues also targeted at 9.2, so let me know if there's anything left you see as a blocker 👌

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/340522)
* Commit: a3a488fd3fa6a1871acdf97705080e37b4767853

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/03f47de4412ec686ef5a09e035949639776c6489) | [after](https://github.com/elastic/kibana/commit/a3a488fd3fa6a1871acdf97705080e37b4767853) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 660 | 676 | +16 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/03f47de4412ec686ef5a09e035949639776c6489) | [after](https://github.com/elastic/kibana/commit/a3a488fd3fa6a1871acdf97705080e37b4767853) | diff |
| --- | --- | --- | --- |
| `@kbn/streamlang` | 160 | 189 | +29 |

#### Any counts in public APIs
> Total count of every any typed public API. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats any` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/03f47de4412ec686ef5a09e035949639776c6489) | [after](https://github.com/elastic/kibana/commit/a3a488fd3fa6a1871acdf97705080e37b4767853) | diff |
| --- | --- | --- | --- |
| `@kbn/streamlang` | 11 | 12 | +1 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/03f47de4412ec686ef5a09e035949639776c6489) | [after](https://github.com/elastic/kibana/commit/a3a488fd3fa6a1871acdf97705080e37b4767853) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 693.0KB | 709.7KB | +16.7KB |

#### Public APIs missing exports
> Total count of every type that is part of your API that should be exported but is not. This will cause broken links in the API documentation system. Target amount is 0. `Run node scripts/build_api_docs --plugin [yourplugin] --stats exports` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/03f47de4412ec686ef5a09e035949639776c6489) | [after](https://github.com/elastic/kibana/commit/a3a488fd3fa6a1871acdf97705080e37b4767853) | diff |
| --- | --- | --- | --- |
| `@kbn/streamlang` | 2 | 3 | +1 |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/03f47de4412ec686ef5a09e035949639776c6489) | [after](https://github.com/elastic/kibana/commit/a3a488fd3fa6a1871acdf97705080e37b4767853) | diff |
| --- | --- | --- | --- |
| `@kbn/streamlang` | 173 | 202 | +29 |

#### ESLint disabled in files
| id | [before](https://github.com/elastic/kibana/commit/03f47de4412ec686ef5a09e035949639776c6489) | [after](https://github.com/elastic/kibana/commit/a3a488fd3fa6a1871acdf97705080e37b4767853) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 12 | 11 | -1 |

#### ESLint disabled line counts
| id | [before](https://github.com/elastic/kibana/commit/03f47de4412ec686ef5a09e035949639776c6489) | [after](https://github.com/elastic/kibana/commit/a3a488fd3fa6a1871acdf97705080e37b4767853) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 10 | 12 | +2 |

#### Total ESLint disabled count
| id | [before](https://github.com/elastic/kibana/commit/03f47de4412ec686ef5a09e035949639776c6489) | [after](https://github.com/elastic/kibana/commit/a3a488fd3fa6a1871acdf97705080e37b4767853) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 22 | 23 | +1 |


</details>


### History
* :green_heart: [Build #339670](https://buildkite.com/elastic/kibana-pull-request/builds/339670) succeeded cc9e397331db4e8796ec91563f4c68da238ae298
* :broken_heart: [Build #339288](https://buildkite.com/elastic/kibana-pull-request/builds/339288) failed fb34199244a6e555239c2f2945f53a2df9c07675
* :broken_heart: [Build #339001](https://buildkite.com/elastic/kibana-pull-request/builds/339001) failed 788f10755d6564f17123a29105fecbff1f62f6f3
* :broken_heart: [Build #338995](https://buildkite.com/elastic/kibana-pull-request/builds/338995) failed a287ae46822fd92112517927c178e74b23c824a8
* :broken_heart: [Build #338953](https://buildkite.com/elastic/kibana-pull-request/builds/338953) failed 773a0f30f0781ce5a3a4f0bce053bda8011dc38d
* :broken_heart: [Build #338561](https://buildkite.com/elastic/kibana-pull-request/builds/338561) failed daf181139272ea0103c3927ef9439b11998846c6

cc @Kerry350

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/340522","number":340522,"commit":"a3a488fd3fa6a1871acdf97705080e37b4767853"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339670","number":339670,"commit":"cc9e397331db4e8796ec91563f4c68da238ae298"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339288","number":339288,"commit":"fb34199244a6e555239c2f2945f53a2df9c07675"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339001","number":339001,"commit":"788f10755d6564f17123a29105fecbff1f62f6f3"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/338995","number":338995,"commit":"a287ae46822fd92112517927c178e74b23c824a8"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/338953","number":338953,"commit":"773a0f30f0781ce5a3a4f0bce053bda8011dc38d"}],"number":340522}
buildkite-pr-comment-->

tonyghiani reviewed (approved):
It looks good to me, thanks for this incredible piece of work! 👏 
The follow-ups are defined, please double-check with @flash1293 if he has any other concern/comment before merging, as we reviewed this in parallel.

florent-leborgne reviewed (approved):
Copy/API docs LGTM!
